### PR TITLE
feat(cart): add cart item in-memory driver and backend

### PR DESCRIPTION
**Pull request by griest024**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Fixes: N/A

## Does this PR introduce a breaking change?
```
[ ] Yes
[ ] No
```

## Other information

### CODE REVIEW

File: libs/cart/testing/src/drivers/in-memory/cart-item/cart-item.service.spec.ts [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';
+
+import { DaffCart, DaffCartItem } from '@daffodil/cart';
+
+import { DaffCartFactory } from '../../../factories/cart.factory';
+import { DaffInMemoryCartItemService } from './cart-item.service';
+import { DaffCartItemFactory } from '@daffodil/cart/testing';
+import { DaffCartItemInput } from 'dist/cart/models/cart-item-input';
+
+describe('Driver | In Memory | Cart | CartItemService', () => {
```
Comment (lderrickable) [inline libs/cart/testing/src/drivers/in-memory/cart-item/cart-item.service.spec.ts]: I think these tests are a little misleading, and I'm guessing we probably do it in the other in memory drivers. But really, the only thing this should test is that calling a method starts a request. The part about actually returning what is supposed to be returned isn't actually done here. So I think the tests should just be like:
```
describe('list', () => {
    it('should send a get request for a list of cart items', done => {
      cartItemService.list(cartId).subscribe(res => {
        done();
      });

      const req = httpMock.expectOne(`${cartItemService.url}/${cartId}/`);

      expect(req.request.method).toBe('GET');
      req.flush();                           <------I'm not sure whether you'll need to flush with actual data or not.
    });
  });
```

File: libs/cart/testing/src/in-memory-backend/cart-items/cart-items.service.spec.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+    beforeEach(() => {
+      reqInfoStub.url = cartUrl;
+      reqInfoStub.req.body = mockCartItemInput;
+      productId = mockCartItemInput.productId;
+
+      result = service.post(reqInfoStub);
+    });
+
+    it('should return a cart item with the product id', () => {
```
Comment (lderrickable) [inline libs/cart/testing/src/in-memory-backend/cart-items/cart-items.service.spec.ts]: should be "return a cart with the added item"

File: libs/cart/testing/src/drivers/in-memory/cart-item/cart-item.service.spec.ts (hunk 1 above)
Comment (griest024) [reply libs/cart/testing/src/drivers/in-memory/cart-item/cart-item.service.spec.ts]: Changed

File: libs/cart/testing/src/in-memory-backend/cart-items/cart-items.service.spec.ts (hunk 2 above)
Comment (griest024) [reply libs/cart/testing/src/in-memory-backend/cart-items/cart-items.service.spec.ts]: Changed

**Comment by lderrickable** [comment]
I think it might make more sense to only have the root-cart service extend the InMemoryService, because only it needs to createDb and do the InMemory stuff. The cart and cart-item inmemory service really only need to be a set of pure functions that mutate the cart object in the root-cart service. What are your thoughts on this?

But I think we should push this PR as it stands, finish the other mock http calls for the other cart drivers, then we'll have a better picture for how this is going to work by testing it in demo. And we can make the other changes then.

**Comment by griest024** [comment]
@lderrickable hm thats a good point, I guess the child backends don't use any of the `InMemoryDbService` features. Additionally, they can't be used on their own so its actually probably a bad idea to have them implement `InMemoryDbService`.

We can push this in its current state. I'll open PRs for the drivers.

**Pull request merged**